// @react-three/drei/core/Stage.js
import _objectWithoutPropertiesLoose from '@babel/runtime/helpers/esm/objectWithoutPropertiesLoose';
import { useRef, useState, useLayoutEffect, createElement } from 'react';
import { Box3, Vector3, Sphere } from 'three';
import { useThree } from '@react-three/fiber';
import { Environment } from './Environment.js';
import { ContactShadows } from './ContactShadows.js';

function Stage(_ref) {
  let {
    children,
    controls,
    contactShadow,
    shadows,
    adjustCamera,
    environment,
    intensity = 1
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, ["children", "controls", "contactShadow", "shadows", "adjustCamera", "environment", "intensity"]);

  const camera = useThree(state => state.camera);
  const outer = useRef(null);
  const inner = useRef(null);
  const [{
    radius,
    width,
    height
  }, set] = useState({
    radius: 0,
    width: 0,
    height: 0
  });
  useLayoutEffect(() => {
    outer.current.position.set(0, 0, 0);
    outer.current.updateWorldMatrix(true, true);
    const box3 = new Box3().setFromObject(inner.current);
    const center = new Vector3();
    const sphere = new Sphere();
    const height = box3.max.y - box3.min.y;
    const width = box3.max.x - box3.min.x;
    box3.getCenter(center);
    box3.getBoundingSphere(sphere);
    set({
      radius: sphere.radius,
      width,
      height
    });
    outer.current.position.set(-center.x, -center.y + height / 2, -center.z);
  }, [children]);
  useLayoutEffect(() => {
    if (adjustCamera) {
      const y = radius / (height > width ? 1.5 : 2.5);
      camera.position.set(0, radius * 1.5, radius * 2.5);
      camera.near = 0.1;
      camera.far = Math.max(5000, radius * 4);
      camera.lookAt(0, y, 0);

      if (controls && controls.current) {
        controls.current.target.set(0, y, 0);
        controls.current.update();
      }
    }
  }, [radius, height, width, adjustCamera]);
  return /*#__PURE__*/createElement("group", props, /*#__PURE__*/createElement("group", {
    ref: outer
  }, /*#__PURE__*/createElement("group", {
    ref: inner
  }, children)), contactShadow && /*#__PURE__*/createElement(ContactShadows, {
    "rotation-x": Math.PI / 2,
    opacity: 0.5,
    width: radius * 2,
    height: radius * 2,
    blur: 2,
    far: radius / 2
  }), /*#__PURE__*/createElement("spotLight", {
    position: [radius, radius * 2, radius],
    intensity: intensity * 2,
    castShadow: shadows
  }), /*#__PURE__*/createElement("pointLight", {
    position: [-radius * 2, -radius * 2, -radius * 2],
    intensity: intensity
  }), environment && /*#__PURE__*/createElement(Environment, {
    preset: environment
  }));
}

export { Stage };
